### DISCUSSION TITLE
Custom (de)serializers

### DISCUSSION THREAD

**Discussion by earthboundkid** [earthboundkid/requests#89] [Ideas] [answered]
I would be nice to be able to support alternative JSON libraries, like #87 and #88. JSON is just one kind of serialization. There's also XML, a la #79 / #82, and presumably other options like binary (Protobuf?) and whatnot.

What if we add a new API like

```go
type Serializer = func(v any) ([]byte, error)
type Deserializer = func([]byte, v any) error

func (rb *Builder) Serialize(s Serializer, v any)
func (rb *Builder) Deserialize(d Deserializer, v any)

// Used if s or d are nil
var (
    DefaultSerializer = json.Marshal
    DefaultDeserializer = json.Unmarshal
)

// Usage
func init() {
  requests.DefaultSerializer = gojson.Marshal
  requests.DefaultDeserializer = gojson.Unmarshal
}

err := requests.
  URL(url).
  Serialize(nil, req).
  Deserialize(nil, res).
  Fetch(ctx)
```

@EvilBorsch, what do you think of this idea?

**Comment by earthboundkid** [comment]
It should probably also have a string param for (default) content type.

**Reply by earthboundkid** [reply]
Having a default content type would also help with the existing issue of application/json getting added whether you want it or not. It could just be a setting that you can toggle.

**Comment by EvilBorsch** [comment]
I think it looks good architecturally, but bad from a ux point of view. It is convenient and understandable for developers to use ToJSON,  if server return for example xml, most convient way is to find ToXML, not to call Serialize with parameters. 

It seems that for complete happiness it is only necessary to give the opportunity to speed up toJSON method (keep in mind that 90%+ usecases will return JSON). Moreover, it seems to me, that this should not be within the framework of conditional FastToJSON, but in the form of a GLOBAL override of the serializer/deserializer for json especially. Because memorize that in each request we need to override serialize is quite hard.

**Reply by EvilBorsch** [reply]
So i will prefer simple API for developers with simple methods to speed up some processes, instead of good from architectural point of view solution, that not obvious (using ToJSON is obvious, knowledge of "Serialize will serialize request in JSON and you can choose serializer" is not obvious). It could work if "Serialize" was on start, and we doesnt have ToJSON. But you need to preserve ToJSON for backward compatibility, so i think best solution is to work with what we already have

**Reply by EvilBorsch** [reply]
And for us especially, we use your library in production in many services, so we want to improve performance without rewriting all our connectors with new fancy API

**Reply by earthboundkid** [reply]
It's a tricky question. On the one hand, you'd want to be able to just set it globally and get the new behavior everywhere. OTOH, you'd also want to be able to override a particular Builder's ToJSON method if a particular endpoint can't work with some serializer for some reason, instead of having to swap everything. Maybe something like this:

```go
type Serializer = func(v any) ([]byte, error)
type Deserializer = func([]byte, v any) error

func (rb *Builder) Serialize(s Serializer, ct string, v any)
func (rb *Builder) Deserialize(d Deserializer, ct string, v any)

// Used by BodyJSON and ToJSON
var (
    JSONSerializer = json.Marshal
    JSONDeserializer = json.Unmarshal
)

func BodyJSON(v any) Body {
   // Something like this
   return Serialize(JSONSerializer, "application/json", v)
}

// Usage
func init() {
  requests.JSONSerializer = gojson.Marshal
  requests.JSONDeserializer = gojson.Unmarshal
}

// One particular request which needs a different serializer
err := requests.
  URL(url).
  Serialize(json.Marshal, "application/json", req).
  Deserialize(nil, res).
  Fetch(ctx)
```

**Reply by EvilBorsch** [reply]
Im agree with you, but i think its a good idea to have in library this two options together, so you can use another json serializer/deserializer globally and use `ToJSON` with this global serializer and overwrite this behaviour, using this brand new `Serialize` method if you want to.

**Reply by EvilBorsch** [reply]
Anyway i will be happy if at least your variant with Serialize/Deserialize will go to master. Because gojson trully improve serialization/deserialization, and we can use it on our high rps interaction between services :)

**Reply by earthboundkid** [reply]
I’ve been busy but I might have time to work on this next week.

**Reply by earthboundkid** [reply, marked as answer]
I started working on #91.

**Reply by EvilBorsch** [reply]
Thank you for your work!
